clients/src/main/java/org/apache/kafka/clients/producer/internals/ProduceRequestResult.java

```diff
@@ -32,21 +34,31 @@
     private final CountDownLatch latch = new CountDownLatch(1);
     private volatile TopicPartition topicPartition;
     private volatile long baseOffset = -1L;
+    private volatile long responseTimestamp = Record.NO_TIMESTAMP;
```

Comment:
This class is initially created by a user thread (the same one that creates the containing`RecordBatch`) and the fields are later updated by the `Sender` thread when the response is received.

Even though it may be OK to rely on the latch for visibility guarantees, we would be updating the fields from a different thread without a happens before edge (i.e. it would be a racy write). I'd rather keep it easier to reason about unless we figure out that this is a concurrency bottleneck. As an aside, it looks like we have a concurrency bug in the producer already: https://issues.apache.org/jira/browse/KAFKA-4557.